src/common/Edelstein.Common.Gameplay.Game/Combat/DamageCalculator.cs

```diff
@@ -0,0 +1,136 @@
+    public uint InitSeed3 { get; }
+    
+    public void Skip()
+        => _rndGenForCharacter.Next(new uint[RndSize]);
+    
+    public async Task<IUserAttackDamage[]> CalculatePDamage(IFieldUserStats attacker, IFieldMobStats target, IUserAttack attack)
+    {
+        var random = new Rotational<uint>(new uint[RndSize]);
+        var skill = attack.SkillID > 0 ? await _skills.Retrieve(attack.SkillID) : null;
+        var skillLevel = skill?.Levels[attack.SkillLevel];
+        var attackCount = skillLevel?.AttackCount ?? 1;
+        var result = new IUserAttackDamage[attackCount];
+
+        _rndGenForCharacter.Next(random.Array);
+        
+        var sqrtACC = Math.Sqrt(attacker.PACC);
+        var sqrtEVA = Math.Sqrt(target.EVA);
+        var hitRate = sqrtACC - sqrtEVA + 100 + attacker.Ar * (sqrtACC - sqrtEVA + 100) / 100;
+
+        hitRate = Math.Min(hitRate, 100);
+
+        if (target.Level > attacker.Level)
+            hitRate -= 5 * (target.Level - attacker.Level);
+        
+        
+        for (var i = 0; i < attackCount; i++)
+        {
+            random.Skip();
+
+            if (hitRate < GetRandomInRange(random.Next(), 0, 100))
+            {
+                result[i] = new UserAttackDamage(0);
+                continue;
+            }
+
+            var damage = GetRandomInRange(random.Next(), attacker.DamageMin, attacker.DamageMax);
+            var critical = false;
+
+            if (target.Level > attacker.Level)
+                damage *= (100d - (target.Level - attacker.Level)) / 100d;
+
+            damage *= (100d - (target.PDR * attacker.IMDr / -100 + target.PDR)) / 100d;
+
+            if (skillLevel != null)
+                damage *= skillLevel.Damage / 100d;
+
+            if (attacker.Cr > 0 && GetRandomInRange(random.Next(), 0, 100) <= attacker.Cr)
+            {
+                var cd = (int)GetRandomInRange(random.Next(), attacker.CDMin, attacker.CDMax) / 100d;
+
+                damage += (int)damage * cd;
+                critical = true;
+            }
+
+            damage = Math.Min(Math.Max(damage, 1), 999999);
+            result[i] = new UserAttackDamage((int)damage, critical);
+        }
+        
+        return result;
+    }
+    
+    public async Task<IUserAttackDamage[]> CalculateMDamage(IFieldUserStats attacker, IFieldMobStats target, IUserAttack attack)    
+    {
+        var random = new Rotational<uint>(new uint[RndSize]);
+        var skill = attack.SkillID > 0 ? await _skills.Retrieve(attack.SkillID) : null;
+        var skillLevel = skill?.Levels[attack.SkillLevel];
+        var attackCount = skillLevel?.AttackCount ?? 1;
+        var result = new IUserAttackDamage[attackCount];
+
+        _rndGenForCharacter.Next(random.Array);
+        
+        for (var i = 0; i < attackCount; i++)
+            result[i] = new UserAttackDamage(0);
+
+        return result;
+    }
+    
+    private double GetRandomInRange(uint rand, double f0, double f1)
+    {
+        if (f1 != f0)
```

Comment:
Cant we just invert the logic? Also wouldn't it benefit to use the Equals overload for double?

        if (f1 == f0) return f0;
        if (f1.Equals(f0)) return f0;